File: HeadFirstDesign_Java/src/Observer/Observer.java
package Observer;

import Observer.impls.*;

/*观察者模式
 * 定义了对象之间的一对多依赖，这样一来，当一个对象改变状态时，
 * 它的所有依赖者都会受到通知并自动更新
 */

/*松耦合
 * 当两个对象之间松耦合，它们依然可以交互，但是不太清楚彼此的细节。
 * 观察者模式提供了一种对象设计，让主题和观察者之间松耦合。
*/
public class Observer
{
    public static void main(String[] args)
    {
        WeatherDataSubject subject=new WeatherDataSubject();//内部有抽象Observer接口列表，会遍历调用update方法
        CurrentConditionsDisplay observer01=new CurrentConditionsDisplay(subject);

        ConcreteConditionsDisplay observer02=new ConcreteConditionsDisplay();
        subject.registerObserver(observer02);

        ForecastDisplay observer03=new ForecastDisplay();
        subject.registerObserver(observer03);

        StatisticsDispaly observer04=new StatisticsDispaly();
        subject.registerObserver(observer04);

        subject.setChange(true);
        subject.setMeasurements(3.14,5,20);
        subject.setMeasurements(5,15, 18);
        subject.setMeasurements(6, 8, 18);
    }
}

File: HeadFirstDesign_CSharp/02Observer/Observer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace _02Observer
{
    //抽象观察者
    interface Observer
    {
        void update(double temp,double humidity,double pressure);//更新
    }
    //抽象显示
    interface DisplayElement
    {
        void display();
    }

    //条件面板
    class ConcreteConditionsDisplay:Observer, DisplayElement
    {
        public void update(double temp, double humidity, double pressure) { }
        public void display() {//显示当前观测值
        }
    }

    //统计面板
    class StatisticsDispaly : Observer, DisplayElement
    {
        public void update(double temp, double humidity, double pressure) { }
        public void display()//显示当前观测值
        {
        }
    }
    //当前状况布告板
    class CurrentConditionsDisplay : Observer, DisplayElement
    {
        private double temperature;
        private double humidity;
        private Subject weatherData;

        public CurrentConditionsDisplay(Subject weatherData)
        {
            this.weatherData = weatherData;
            weatherData.registerObserver(this);
        }
        public void update(double temp, double humidity, double pressure) {
            temperature = temp;
            this.humidity = humidity;
            display();
        }
        public void display()//基于观测值的其他内容
        {
            Console.WriteLine("当前温度："+ temperature+" 当前湿度："+ humidity);
        }
    }
    //预报面板
    class Forecast : Observer, DisplayElement
    {
        public void update(double temp, double humidity, double pressure) { }
        public void display()//显示天气预报
        {
        }
    }
}
